x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java

```diff
@@ -0,0 +1,534 @@
+                            var provider = new InferenceProvider(
+                                service.get(),
+                                service.get()
+                                    .parsePersistedConfigWithSecrets(
+                                        inferenceId,
+                                        unparsedModel.taskType(),
+                                        unparsedModel.settings(),
+                                        unparsedModel.secrets()
+                                    )
+                            );
+                            executeShardBulkInferenceAsync(inferenceId, provider, requests, onFinish);
+                        } else {
+                            try (onFinish) {
+                                for (int i = 0; i < requests.size(); i++) {
+                                    var request = requests.get(i);
+                                    inferenceResults.get(request.index).failures.add(
+                                        new ResourceNotFoundException(
+                                            "Inference service [{}] not found for field [{}]",
+                                            unparsedModel.service(),
+                                            request.field
+                                        )
+                                    );
+                                }
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception exc) {
+                        try (onFinish) {
+                            for (int i = 0; i < requests.size(); i++) {
+                                var request = requests.get(i);
+                                inferenceResults.get(request.index).failures.add(
+                                    new ResourceNotFoundException("Inference id [{}] not found for field [{}]", inferenceId, request.field)
+                                );
+                            }
+                        }
+                    }
+                };
+                modelRegistry.getModelWithSecrets(inferenceId, modelLoadingListener);
+                return;
+            }
+            int currentBatchSize = Math.min(requests.size(), batchSize);
+            final List<FieldInferenceRequest> currentBatch = requests.subList(0, currentBatchSize);
+            final List<FieldInferenceRequest> nextBatch = requests.subList(currentBatchSize, requests.size());
+            final List<String> inputs = currentBatch.stream().map(FieldInferenceRequest::input).collect(Collectors.toList());
+            ActionListener<List<ChunkedInferenceServiceResults>> completionListener = new ActionListener<>() {
+                @Override
+                public void onResponse(List<ChunkedInferenceServiceResults> results) {
+                    try {
+                        for (int i = 0; i < results.size(); i++) {
```

Comment:
My point is that this loop depends on the `results` list being an `ArrayList` to be performant. Given that we get this list from service external to this class, how do we guarantee that?

I'm not suggesting that we hard-code the implementation, that would make the code too brittle. But it may be a good idea to use iterators to advance through the lists so that performance is not dependent on the list implementation.